File: 백준/Gold/12912. 트리 수정/트리 수정.java
import java.util.*;
import java.io.*;

public class Main {

    static int max_node = -1;
    static long max_dis;
    static int N, temp_cur, temp_node, temp_dis;
    static long[] dist;
    static boolean[] v;
    static int[][] edge;
    static ArrayList<int[]>[] graph;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        N = Integer.parseInt(br.readLine());
        edge = new int[N][3];
        dist = new long[N];
        graph = new ArrayList[N];

        for(int i = 0; i < N; i ++) graph[i] = new ArrayList<>();
        for(int i = 0; i < N - 1; i ++){
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());

            edge[i] = new int[] { from, to, cost };
            graph[from].add(new int[] { to, cost });
            graph[to].add(new int[] { from, cost });
        }

        long answer = 0;
        for(int i = 0; i < edge.length - 1; i ++){

            int t1 = edge[i][0];
            int t2 = edge[i][1];
            int cost = edge[i][2];

            // 입력으로 주어진 간선 중 하나를 잘라 별도의 트리로 만듦
            v = new boolean[N];
            int node1 = dfs(t1, edge[i]);

            v = new boolean[N];
            dfs(node1, edge[i]);

            v = new boolean[N];
            int node2 = dfs(t2, edge[i]);

            v = new boolean[N];
            dfs(node2, edge[i]);

            answer = Math.max(answer, dist[node1] + dist[node2] + cost);
        }
        System.out.println(answer);
    }

    // 현재 node에서 가장 멀리있는(가중치가 큰) 노드 탐색
    public static int dfs(int node, int[] edge){
        v[node] = true;
        int lastNode = node;
        dist[node] = 0;
        for(int[] next : graph[node]){
            if ((node == edge[0] && next[0] == edge[1]) || (node == edge[1] && next[0] == edge[0])) continue;
                if(v[next[0]]) continue;
                v[next[0]] = true;
                int cur = dfs(next[0], edge);
                if(dist[node] < dist[next[0]] + next[1]){
                    dist[node] = dist[next[0]] + next[1];
                    lastNode = cur;
                }
        }

        return lastNode;
    }
}

File: 백준/Gold/12912. 트리 수정/트리 수정.py
import sys
sys.setrecursionlimit(100000)
input = sys.stdin.readline

N = int(input().strip())
nodes, edges, v, dist = [[] for _ in range(N)], [], [False] * N, [0] * N

def dfs(node, e1, e2):
    cur, dist[node] = node, 0
    for nxt in nodes[node]:
        nxt_node, nxt_cost = nxt
        if nxt_node == node or v[nxt_node]:
            continue

        if (node == e1 and nxt_node == e2) or (node == e2 and nxt_node == e1):
            continue
        
        v[nxt_node] = True
        tmp = dfs(nxt_node, e1, e2)
        if dist[nxt_node] + nxt_cost > dist[node]:
            dist[node] = dist[nxt_node] + nxt_cost
            cur = tmp

    return cur

for _ in range(N-1):
    s, e, c = map(int, input().strip().split())
    nodes[s].append((e, c))
    nodes[e].append((s, c))

    edges.append((s, e, c))

ans = 0
for edge in edges:
    e1, e2, c = edge

    # 분리된 1번 트리의 지름 구하기
    v = [False] * N
    v[e1] = True
    n1 = dfs(e1, e1, e2)

    v = [False] * N
    v[n1] = True
    n2 = dfs(n1, e1, e2)

    # 분리된 2번 트리의 지름 구하기
    v = [False] * N
    v[e2] = True
    n3 = dfs(e2, e1, e2)

    v = [False] * N
    v[n3] = True
    n4 = dfs(n3, e1, e2)

    ans = max(ans, dist[n1] + dist[n3] + c)

print(ans)